pytest_bdd/scenario.py

```diff
@@ -10,47 +10,92 @@
     scenario_name="Publishing the article",
 )
 """
-import collections
 import os
 import re
+from collections import namedtuple
+from itertools import chain, product
+from operator import attrgetter, itemgetter
 
 import pytest
 from _pytest.fixtures import FixtureLookupError
 
 from . import exceptions
 from .feature import get_feature, get_features
+from .parser import STEP_PARAM_RE, STEP_PARAM_TEMPLATE
 from .steps import get_step_fixture_name, inject_fixture
 from .utils import CONFIG_STACK, get_args, get_caller_module_locals, get_caller_module_path
 
 PYTHON_REPLACE_REGEX = re.compile(r"\W")
 ALPHA_REGEX = re.compile(r"^\d+_*")
 
 
+def generate_partial_substituted_step_parameters(name: str, request):
+    """Returns step name with substituted parameters from fixtures, example tables, param marks"""
+    yield name
+    matches = re.finditer(STEP_PARAM_RE, name)
+    for match in matches:
+        param_name = match.group(1)
+        name = re.sub(
+            STEP_PARAM_TEMPLATE.format(param=re.escape(param_name)),
+            str(request.getfixturevalue(param_name)),
+            name,
+            count=1,
+        )
+        yield from generate_partial_substituted_step_parameters(name, request)
+
+
+ParserFixturedefFunctionMapping = namedtuple("ParserFixturedefFunctionMapping", ["parser", "fixturedef_function"])
+
+StepNameParserFixturedefFunctionMapping = namedtuple(
+    "StepNameParserFixturedefFunctionMapping", ["step_name", "parser_fixturedef_function"]
+)
+
+
 def find_argumented_step_fixture_name(name, type_, fixturemanager, request=None):
     """Find argumented step fixture name."""
     # happens to be that _arg2fixturedefs is changed during the iteration so we use a copy
-    for fixturename, fixturedefs in list(fixturemanager._arg2fixturedefs.items()):
-        for fixturedef in fixturedefs:
-            parser = getattr(fixturedef.func, "parser", None)
-            if parser is None:
-                continue
-            match = parser.is_matching(name)
-            if not match:
-                continue
 
-            converters = getattr(fixturedef.func, "converters", {})
-            for arg, value in parser.parse_arguments(name).items():
-                if arg in converters:
-                    value = converters[arg](value)
-                if request:
-                    inject_fixture(request, arg, value)
-            parser_name = get_step_fixture_name(parser.name, type_)
+    parsers_fixturedef_functions = filter(
```

Comment:
Can this use list comprehension syntax? I think it might be a little bit more readable if we don't use map/filter with lambdas.

Reply:
Done